OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs

```diff
@@ -121,6 +122,24 @@ bool ShouldLandAtBuilding(Actor self, Actor dest)
 
 		public override Activity Tick(Actor self)
 		{
+			if (ChildActivity != null)
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+
+				// restartAfterChildsFinished serves the purpose of avoiding queueing RTB as ChildActivity of itself
+				// in cases where we want RTB to trigger again after having finished returning to a resupplier
+				// that was reserved when the return was originally triggered.
+				if (ChildActivity != null || restartAfterChildsFinished)
```

Comment:
I'm confused. How does RTB queue itself as a `ChildActivity`?

Reply:
Ok, that comment is a bit hard to read (I just wasn't sure how to expess it without making the comment even longer).

What I mean is: Previously, for example https://github.com/OpenRA/OpenRA/pull/16241/files#diff-1f1f9ac8abeec88534accb714c8fbb7bL141 would queue `this` after some other activities, to make the aircraft return to base first and *then* check for any free resuppliers (if there weren't any before).
But queueing `this` feels risky, I don't want to run into an infinite loop. So instead, I use this boolean to basically 'restart' `ReturnToBase` "from the beginning" after the child activities have finished, provided `dest` was null last time we checked and we're still looking for an available resupplier.

Reply:
Queuing `this` as a child activity of itself is indeed bogus.

I still don't understand what's going on here.  Is the goal to skip the `abortOnResupply` cancellation?
Why can't we just reset `dest` to null if we want to force a new search?

IMO the first step should be to rewrite this first block as something like
```csharp
if (ChildActivity != null)
{
	ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
	if (ChildActivity != null)
		return this;
}

// resupplied is a new flag set when ResupplyAircraft is queued
if (resupplied && abortOnResupply)
    Cancel(self);
```
so that the child handling follows the same pattern as other activitys, allowing it to be deduplicated.  The next steps after that might become clearer then.

Reply:
As far as I can tell, what you're trying to do is already the default. You don't need anything more than the standard `if (ChildActivity != null)` block. Activities queueing `this` after some other activities is (was) by no means an uncommon pattern and it is exactly this pattern that the `ChildActivity` based approach is supposed to replace. Have a look at the code changes in #16348. Is the situation here much different?

Reply:
@pchote No, the goal is simply to make the activity 'reset' after the child activities finish **if** dest is/was a *reserved* resupplier (because there weren't any unreserved ones when we last checked), and make the activity end if `dest` was not reserved (so we should be done resupplying when the child activities finish).

@tovl The problem here is that with the standard `if (ChildActivity != null)` block without anything else, we'd be lacking a `return NextActivity` for the case of "last search for a resupplier succeeded and we're done resupplying now, so this activity should end now", as the only other `return NextActivity` (currently) triggers on cancellation only.

Reply:
But yes, @pchote's example already does the trick. I just had to add `resupplied` to one of the `if (...) -> return NextActivity` blocks.

Reply:
Ah, now I get it. What you have right now looks like it should do the trick.